scipy/sparse/sparsetools/util_par_config.h

```diff
@@ -0,0 +1,64 @@
+#ifndef __SPTOOLS_UTIL_PAR_H__
+#define __SPTOOLS_UTIL_PAR_H__
+
+#include "util_par_algorithms.h"
+
+using util_par::for_each_chunk;
+using util_par::iota_iter;
+
+class worker_config {
+private:
+    // Number of worker threads. 1 means sequential, 0 means number of CPUs.
+    int num_workers = 1;
+//    int num_workers = std::min(16, std::thread::hardware_concurrency());
```

Comment:
Not quite. NumPy, SciPy, scikit-learn are all single-threaded by default, and SciPy/scikit-learn have keywords to switch to multiple threads. The only exception is BLAS/LAPACK calls, and that's because it's not possible to control at runtime in a reasonable manner (and exact threading depends on build settings and env vars). I've been in some discussions with pandas devs as well, where the majority opinion was that single-threaded is the correct default. If you control the whole stack with your own framework (e.g., PyTorch/TF/JAX), it's a much easier story, but that doesn't apply here.

Using all threads by default seems logical at first sight, but you find soon enough that it breaks things when there's higher-level parallelism (e.g., Dask, multiprocessing).

It'd be great to have better coordination on parallelism control, but it's tough. See https://thomasjpfan.github.io/parallelism-python-libraries-design/ for a good summary of where we are at and what the options are to improve.

Reply:
Apologies for the late response!

> NumPy, SciPy, scikit-learn are all single-threaded by default

I think this is true on the python code level, but in practice when I run something like scikit learns dense PCA I will be using many threads by default, since matrix multiplication is multithreaded by default with pypi and conda numpy distributions.

> The only exception is BLAS/LAPACK calls, and that's because it's not possible to control at runtime in a reasonable manner

Scikitlearn, via `joblib` and `threadpoolctl` do control the threading of BLAS/ LAPACK calls to avoid over subscription when their higher level parallelism is used.

> I've been in some discussions with pandas devs as well, where the majority opinion was that single-threaded is the correct default. 

I would be curious how this interacts with their increasing use of arrow via pyarrow, since pyarrow's compute and IO is [multithreaded by default](https://arrow.apache.org/docs/python/generated/pyarrow.cpu_count.html).

> Using all threads by default seems logical at first sight, but you find soon enough that it breaks things when there's higher-level parallelism (e.g., Dask, multiprocessing).

If we are requiring some users to have to think more about parallelism, I think it's more reasonable to have those users be the ones using a higher level parallelism. I believe so much code is highly multithreaded by default in the scientific python ecosystem (`pyarrow`, `polars`, many things using `numba`, `numpy`'s dot product, etc.) that I think people who want to use higher level parallelism already need to take steps to mitigate oversubscription.

 I think one of the nice things this PR does is hook into some nice tooling for higher level parallelism. E.g. if this code is called within a jobllib Parallel job it will automatically scale the number of threads used via `threadpoolctl` (thanks to @alugowski's upstream improvements!).

Dask also exerts control over the number of BLAS threads each job has access to, though I believe it's a different mechanism.

### User expectations

I think it would be unintuitive if `a @ b` was multithreaded for dense arrays and single threaded for sparse arrays by default.